osidb/models/affect.py

```diff
@@ -473,15 +482,31 @@ def _validate_special_consideration_flaw(self, **kwargs):
                     f"{affected_special_consideration_package} is missing statement.",
                 )
 
-    def _validate_purl(self, **kwargs):
+    def _validate_purl_and_ps_component(self, **kwargs):
         """
-        Checks that the purl field can be parsed into PackageURL object.
+        Validate that purl and ps_component comply with one of the following options:
+        * purl is correct and ps_component is not provided
+        * purl is correct and ps_component matches the one included in purl
         """
-        try:
-            if self.purl:
-                PackageURL.from_string(self.purl)
-        except Exception as exc:
-            raise ValidationError(f"Invalid purl '{self.purl}': {exc}")
+        if not self.purl and not self.ps_component:
+            raise ValidationError(
+                f"Affect ({self.uuid}) for {self.ps_module} must have either purl or ps_component."
+            )
+
+        elif self.purl and not self.ps_component:
+            try:
+                self.ps_component = PackageURL.from_string(self.purl).name
+            except Exception as exc:
+                raise ValidationError(
+                    f"Affect ({self.uuid}) for {self.ps_module} has an invalid purl "
+                    f"'{self.purl}': {exc}. Provide either a valid one or ps_component."
+                )
+
+        elif self.purl and self.ps_component != PackageURL.from_string(self.purl).name:
```

Comment:
Here is still a non-caught exception possibly happening during the PURL parsing which would result in potentially non-user-readable error message out of our control.